[Elastic Charts] Added value labels styles for bar charts

miukimiu opened the pull request:
### Summary

Closes #4267.

This PR adds better value labels default styles for bar charts.

By default the value labels once enabled are going to: 

* Be vertical and horizontal centered. The ideal positions would be based on the rotation of the bars. But this can't be achieved because we can't pass to the theme object what is the current rotation. So vertical and horizontal centered labels seem a good compromise. 
* The color will adapt to black or white according to the background (`textInvertible`)
 
<img width="2472" alt="Frame 5@2x" src="https://user-images.githubusercontent.com/2750668/127162489-643eb814-b016-45be-83e9-1936eef48094.png">

The following demo `#/elastic-charts/categorical` was updated and a "Show value labels" switch was added. The value labels adapt the position according to the configuration. They are positioned close to the edge of the bars.

<img width="1127" alt="Screenshot 2021-09-09 at 17 01 38" src="https://user-images.githubusercontent.com/2750668/132721444-df5aec84-3eb6-4673-ac5f-7bf3d2ca8fb1.png">


When it's multi-series and stacked the value labels are positioned vertically and horizontally in the center.

<img width="1125" alt="Screenshot 2021-09-09 at 17 01 54" src="https://user-images.githubusercontent.com/2750668/132721457-d03db7b4-cb2f-4251-97d5-463527911fee.png">


### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- ~[ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**~
- ~[ ] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**~
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples~
- ~[ ] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**~
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

miukimiu commented on src/themes/charts/themes.ts:
```diff
@@ -112,9 +112,13 @@ function createTheme(colors: any): EuiChartThemeType {
+          fill: {
+            textInvertible: true,
+            textContrast: true,
+            textBorder: 0.5,
```
@markov00  `textBorder: 0` or `textBorder: false` doesn't work.

markov00 replied:
Hey @miukimiu I've created a PR to fix this behaviour: https://github.com/elastic/elastic-charts/pull/1182
with that, you can specify 0 and the `textBorder` will not be visible

miukimiu replied:
Thanks @markov00!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

cchaos commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -166,7 +193,7 @@ export const CategoryChart = () => {
 
         <EuiFlexItem>
-          <MultiChartCard onChange={onMultiChange} />
+          <MultiChartCard onChange={onMultiChange} showValueLabels />
```
I would put the value labels in it's own, new card. This one is specifically labelled `Single vs multiple series` but the labels aren't limited to that scope. 

Then you can take the opportunity to give **guidelines 😉** on when it's good to use direct chart labels.

cchaos replied:
I would also have at least one configuration keep the default centered label.

cchaos reviewed (commented):
👍 💯 Much better defaults for sure!

I have mainly suggestions around documentation and giving more to the consumer than just a toggle.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

markov00 reviewed (commented):
I've only one comment:
when rendering stacked bars, I agree with the choice of having the labels centered in the bars, the value corresponds to the size of the bar and is not directly connected with the Y-axis.
Instead, when you only have a single series, the label could be placed near the top edge of the bar. In that way, you link the label to its corresponding position on the Y-axis. This also limits the eye movements when reading the bar, reducing the gap between the top edge and the value label.
Using your screenshot as an example, I'm noticing that my eyes are focused or on the value labels (centered) or on the bars' shape/trend. Moving the label closer to the edge makes the readability easier 
![image](https://user-images.githubusercontent.com/1421091/127495243-29e9d0ad-8981-40a6-9783-e2aeae1c99cf.png)

monfera commented:
I agree it'd be better for the numbers in the non-stacked case to be at the top. Also, the custom is that if there's room past the single bar (in the positive or negative direction, wherever the bar extends toward) then the label would be outside the bar. There's more text contrast there. [Example](https://depictdatastudio.com/wp-content/uploads/2017/02/EmeryAnalytics_Flipping-Bar-Charts-1.png)

This may not be the place to raise this issue, but our bar charts, and charts in general are configured with a suboptimal balance of font size and data glyph (bar rectangle, etc.) shape. The fonts are small and the data shapes are large. It looks like a functional prototype not the final product. It's also an issue from an accessibility viewpoint: the user gets a huge piece of help in determining the screenspace size of the data shapes, but it's at the expense of font size. Also, even if there's ample room and the user has something like 3 bars, it shouldn't take up all that space. There ought to be a ratio of font size to bar width. For example, on the above images, the bar thickness is easily 3x that of the category labels, and the slightly taller value label inside the bar is also barely over 1/3rd of the height. 1.618x (golden ratio) to around 2x ratio would work but 3x is visually and accessibility wise unbalanced. Big bars, small fonts is the main way to recognize Kibana visualizations in the wild

miukimiu commented:
> Instead, when you only have a single series, the label could be placed near the top edge of the bar. In that way, you link the label to its corresponding position on the Y-axis. This also limits the eye movements when reading the bar, reducing the gap between the top edge and the value label.

@markov00 I agree. But how can I get this as default in the theme? The theme is static and doesn't know what is the current rotation state. Inside the theme object, we don't know if the chart is rotated or if it is stacked. That's why I had to pick one default alignment that would make more sense to most use cases.

Then in the demo, we're showing that you don't need to use the defaults but implement an alignment that makes sense:
<img width="1182" alt="Frame 1@2x" src="https://user-images.githubusercontent.com/2750668/127499958-2a64bba2-a75b-4528-b876-be1101d44d6c.png">

The first time a value label is enabled it is aligned at the end of the bar chart. But then you have the option to center align. The idea is that users can play around with the toggles and copy the configuration that makes sense for their use case. We even recommend just use value labels when it's strictly necessary.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

miukimiu commented:
### Default theme 

@markov00 as we agreed, the default theme keeps the value labels vertically/horizontally centered. 

### Demo (https://eui.elastic.co/pr_4845/#/elastic-charts/categorical)

@cchaos and @markov00 for the demo in our docs, I removed the **center value labels** switch option. Now according to the configuration, the value labels adapt.

> when rendering stacked bars, I agree with the choice of having the labels centered in the bars

✔️  When stacked the labels get centered

<img width="1125" alt="Screenshot 2021-09-09 at 17 01 54" src="https://user-images.githubusercontent.com/2750668/132725185-92322d4f-20e0-4dcc-ba24-0229d10104dc.png">


> Instead, when you only have a single series, the label could be placed near the top edge of the bar.

✔️ They are now positioned near the edge of the bar

<img width="1118" alt="Screenshot 2021-09-09 at 17 15 55" src="https://user-images.githubusercontent.com/2750668/132725214-cc65a444-3ce6-438e-8e00-6e38cd663122.png">


### Generated configuration

I also improved the data of the "Copy code of current configuration" and removed the `orderBy` (lodash) that was included in the generated code.

![Frame 3](https://user-images.githubusercontent.com/2750668/132734511-55a764cf-e4ac-47b0-a6a0-42e41501c6eb.png)

### Next

> Also, the custom is that if there's room past the single bar (in the positive or negative direction, wherever the bar extends toward) then the label would be outside the bar. There's more text contrast there. Example elastic/elastic-charts#1277

@monfera I'm looking forward to https://github.com/elastic/elastic-charts/issues/1277. Render bar labels outside of the bar is going to be a great enhancement. 👍🏽 

@markov00 and @cchaos can you both do another review?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

markov00 commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -70,6 +74,114 @@ export const CategoryChart = () => {
+    barSeriesStyle: {
+      displayValue: {
+        ...theme.barSeriesStyle.displayValue,
+        offsetX: rotated ? 2 : 0,
+        offsetY: rotated ? 0 : -2,
```
From what I see, this is used as padding on the text. We can probably introduce a default or custom padding and we can get rid of this part

markov00 replied:
https://github.com/elastic/elastic-charts/issues/1375

miukimiu replied:
I agree 👍🏽

markov00 commented on src/themes/charts/themes.ts:
```diff
@@ -100,9 +100,17 @@ function createTheme(colors: any): EuiChartThemeType {
       barSeriesStyle: {
         displayValue: {
-          fontSize: 8,
+          fontSize: 10,
```
You can also set a min and max font size, the code automatically uses the right size depending on the available space.

markov00 reviewed (approved):
Looks good to me, tested locally, and works great.

cchaos commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -70,6 +74,114 @@ export const CategoryChart = () => {
+  ? EUI_CHARTS_THEME_DARK.theme
+  : EUI_CHARTS_THEME_LIGHT.theme;
+
+const customTheme = {
+  ...theme,
+  barSeriesStyle: {
+    displayValue: {
+      ...theme.barSeriesStyle.displayValue,
+      offsetX: ${rotated ? '2' : '0'},
+      offsetY: ${rotated ? '0' : '-2'},
+      ${multi && stacked ? defaultAlignmentToCopy : alignmentRotatedToCopy},
+    },
+  },
+};`;
```
I know it's a lot more work, and honestly, I wish there was just an overall better way to handle all this, but can you make this whole extra configuration based on the labels toggle? It's a lot of unnecessary theme changes if consumers aren't showing the value labels.

cchaos commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -70,6 +74,114 @@ export const CategoryChart = () => {
+
+  const textToCopy = `${chartVariablesToCopy}
+
+${removeEmptyLines(chartConfigurationToCopy)}
+  `;
```
```suggestion
${removeEmptyLines(chartConfigurationToCopy)}`;
```

cchaos reviewed (approved):
Overall the new toggle and theme settings LGTM. Had a couple questions and suggestions, but ok to merge in as-is.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4845/

markov00 commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -70,6 +74,114 @@ export const CategoryChart = () => {
+  <Axis
+    id="left-axis"
+    position={${rotated ? '"bottom"' : '"left"'}}
+    ${formatted ? 'tickFormat={d => `${round(Number(d) / 1000, 2)}k`}' : ''}
```
thank you @cchaos you are right, this is an issue in our code. I will check and fix the bug soon.

markov00 replied:
https://github.com/elastic/elastic-charts/issues/1387
